tool/tsh/common/db_exec.go

```diff
@@ -0,0 +1,647 @@
+
+	return trace.Wrap(execCommand.run())
+}
+
+// databaseExecClient is a wrapper of client.TeleportClient that makes backend
+// calls. can be mocked for testing.
+type databaseExecClient interface {
+	close() error
+	getProfileStatus() *client.ProfileStatus
+	getAccessChecker() services.AccessChecker
+	issueCert(context.Context, *databaseInfo) (tls.Certificate, error)
+	listDatabasesWithFilter(context.Context, *proto.ListResourcesRequest) ([]types.Database, error)
+}
+
+type databaseExecCommand struct {
+	cf          *CLIConf
+	tc          *client.TeleportClient
+	client      databaseExecClient
+	makeCommand func(context.Context, *databaseInfo, string, string) (*exec.Cmd, error)
+	summary     databaseExecSummary
+}
+
+func newDatabaseExecCommand(cf *CLIConf) (*databaseExecCommand, error) {
+	if err := checkDatabaseExecInputFlags(cf); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	tc, err := makeClient(cf)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	sharedClient, err := newSharedDatabaseExecClient(cf, tc)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	commandMaker, err := newDatabaseExecCommandMaker(cf.Context, tc, sharedClient.getProfileStatus())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &databaseExecCommand{
+		cf:          cf,
+		tc:          tc,
+		client:      sharedClient,
+		makeCommand: commandMaker.makeCommand,
+	}, nil
+}
+
+func (c *databaseExecCommand) run() error {
+	// Fetch.
+	dbs, err := c.getDatabases()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Execute  in parallel.
+	group, groupCtx := errgroup.WithContext(c.cf.Context)
+	group.SetLimit(c.cf.ParallelJobs)
+	for _, db := range dbs {
+		group.Go(func() error {
+			result := c.exec(groupCtx, db)
+			c.summary.add(result)
+			return nil
+		})
+	}
+
+	// Print summary.
+	defer func() {
+		switch {
+		case c.cf.OutputDir != "":
+			c.summary.printAndSave(c.cf.Stdout(), c.cf.OutputDir)
+		case len(dbs) > 1:
+			c.summary.print(c.cf.Stdout())
+		}
+	}()
+
+	return trace.Wrap(group.Wait())
+}
+
+func (c *databaseExecCommand) close() {
+	if err := c.client.close(); err != nil {
+		logger.WarnContext(c.cf.Context, "Failed to close client", "error", err)
+	}
+}
+
+func checkDatabaseExecInputFlags(cf *CLIConf) error {
+	// Pick an arbitrary number for max connections to avoid flooding the
+	// backend. The limit can be overwritten with the "TELEPORT_PARALLEL_JOBS"
+	// env var.
+	const maxParallelJobs = 10
+	if maxStr := os.Getenv(parallelJobsEnvVar); maxStr != "" {
+		if maxValue, err := strconv.Atoi(maxStr); err != nil || maxValue <= 0 {
+			return trace.BadParameter("environment variable %s must be a positive integer", parallelJobsEnvVar)
+		} else {
+			cf.ParallelJobs = maxValue
+		}
+	} else if cf.ParallelJobs < 1 || cf.ParallelJobs > maxParallelJobs {
+		return trace.BadParameter(`--parallel must be between 1 and %v.
+
+You may use TELEPORT_PARALLEL_JOBS to bypass limit. However, setting a very high
```

Comment:
Use constant instead of env var name verbatim here.

Reply:
Also, what's the purpose of this environment variable compared to the `--parallel` flag? Seems a little strange that you can set the limit with the flag but set a higher value using env var?

Reply:
dropped env var overwrite for this version. we can come back to this if customer complains about the max 10